doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,271 @@
+
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 < the timestamp of T2.
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is write-only, and
+  T2 is read-only.
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 == the timestamp of T2, the timestamp for T1
+  was assigned before the timestamp of T2, T1 is read-only, and T2 is read-only.
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 == the timestamp of T2, the Global ID of T1 is
+  less than the Global ID of T2, T1 is write-only, and T2 is write-only. NOTE: This works because currently you can only
+  write to a single table in a transaction.
+
+### Timelines
+
+Timelines are the unit of consistency that Materialize provides and can be read about in detail in the
+[Timeline design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20210408_timelines.md)
+. Every database object belongs to a single timeline and each read transaction can only include objects from a single
+timeline. Each timeline defines their own notion of time and consistency guarantees are only provided within a single
+timeline. There can only be at most a single Coordinator thread assigning timestamps per timeline (though multiple
+timelines can share a thread if they want). All the properties discussed below need to happen on a per-timeline basis,
+and require no communication across timelines (except to serialize access to the catalog).
+
+Each timeline needs some definition for the current time. For example timelines that track real time can use the system
+clock for the current time. The current time can go backwards, but timestamps assigned must be non-decreasing.
+
+The document also allows user tables to exist in any timeline.
+
+### Global Timestamp
+
+All reads to any object and all writes to user tables are assigned a timestamp by the Coordinator on a single thread.
+The logic for determining what timestamp to return must satisfy:
+
+- If it's a read, then it is larger than or equal to the `since` of all objects involved in the query.
+- If it's a write, then it is larger than or equal to the `upper` of the table.
+- It is larger than or equal to the timestamp of the most recently completed query.
+    - If the query is a write and the most recently completed query was a read, then the timestamp returned must be
+      strictly larger than the previous timestamp.
+
+We can accomplish this by using a single global timestamp per timeline. The timestamp will be initialized to some valid
+initial value and be advanced periodically to some valid value that is higher than the previous value. The periodic
+advancements should maintain the property that the global timestamp is within some error bound of all the timestamps
+being assigned to data from upstream sources. All reads are given a timestamp equal to the global timestamp. All writes
```

Comment:
"All reads are given..." might not meet the `since` criteria from above, unless `since`s are also constrained somehow.

Reply:
Added something to explain why this meets the `since` criteria.